presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizersFactory.java

```diff
@@ -82,6 +83,7 @@ public PlanOptimizersFactory(Metadata metadata, SqlParser sqlParser, FeaturesCon
                 new CountConstantOptimizer(),
                 new WindowFilterPushDown(metadata), // This must run after PredicatePushDown and LimitPushDown so that it squashes any successive filter nodes and limits
                 new MergeProjections(),
+                new ImplementIntersectAsUnion(),
```

Comment:
I can move it further to the top of the list, but I think it is best to place it after SetFlatteningOptimizer rewrite.
After SetFlatteningOptimizer the IntersectNode/s will have all its "sources" grouped, and thus it becomes a lot easier to create the new marker projection columns based on the number of "sources". 

Immediately after DesugaringOptimizer, we may have nested IntersectNodes which will make this rewrite more complex (especially when there are multiple INTERSECTs). What do you think?